pkg/runtime/scheduler/scheduler.go

```diff
@@ -36,183 +42,222 @@ type Options struct {
-		return ctx.Err()
+	var err error
+	s.clients, err = clients.New(ctx, clients.Options{
+		Addresses: addresses,
+		Security:  s.security,
+	})
+	if err != nil {
+		return err
 	}
 
-	if err == io.EOF {
-		log.Warnf("Received EOF, re-establishing connection: %v", err)
-		return nil
+	close(s.readyCh)
+	s.htarget.Ready()
+	s.lock.Unlock()
+
+	err = s.cluster.RunClients(ctx, s.clients)
+
+	s.lock.Lock()
+	s.readyCh = make(chan struct{})
+	s.htarget.NotReady()
+	s.lock.Unlock()
+
+	return err
+}
+
+func (s *Scheduler) Next(ctx context.Context) (schedulerv1pb.SchedulerClient, error) {
+	var client schedulerv1pb.SchedulerClient
+	if err := s.callWhenReady(ctx, func(ctx context.Context) error {
+		var err error
+		client, err = s.clients.Next()
+		return err
+	}); err != nil {
+		return nil, err
 	}
+	return client, nil
+}
 
-	if errors.Is(err, context.Canceled) {
+func (s *Scheduler) All(ctx context.Context) ([]schedulerv1pb.SchedulerClient, error) {
+	var clients []schedulerv1pb.SchedulerClient
+	if err := s.callWhenReady(ctx, func(ctx context.Context) error {
+		clients = s.clients.All()
 		return nil
+	}); err != nil {
+		return nil, err
 	}
+	return clients, nil
+}
 
-	return fmt.Errorf("error watching scheduler jobs: %w", err)
+func (s *Scheduler) StartApp(ctx context.Context) {
+	s.cluster.StartApp()
 }
 
-// watchJobs watches for job triggers from all scheduler clients.
-func (m *Manager) watchJobs(ctx context.Context, appTarget bool, actorTypes []string) error {
-	req := &schedulerv1pb.WatchJobsRequest{
-		WatchJobRequestType: &schedulerv1pb.WatchJobsRequest_Initial{
-			Initial: &schedulerv1pb.WatchJobsRequestInitial{
-				AppId:     m.appID,
-				Namespace: m.namespace,
-			},
-		},
-	}
+func (s *Scheduler) StopApp(ctx context.Context) {
+	s.cluster.StopApp()
+}
 
-	var acceptJobTypes []schedulerv1pb.JobTargetType
-	if appTarget {
-		acceptJobTypes = append(acceptJobTypes, schedulerv1pb.JobTargetType_JOB_TARGET_TYPE_JOB)
-	}
+func (s *Scheduler) callWhenReady(ctx context.Context, fn concurrency.Runner) error {
+	for {
+		s.lock.RLock()
+		readyCh := s.readyCh
+		s.lock.RUnlock()
+
+		select {
+		case <-s.disabled:
+			return errors.New("scheduler not enabled")
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-readyCh:
+		}
 
-	if len(actorTypes) > 0 {
-		acceptJobTypes = append(acceptJobTypes, schedulerv1pb.JobTargetType_JOB_TARGET_TYPE_ACTOR_REMINDER)
-		req.GetInitial().ActorTypes = actorTypes
+		// Ensure still ready as there is a race from above.
+		s.lock.RLock()
+		select {
+		case <-s.readyCh:
+			defer s.lock.RUnlock()
+			return fn(ctx)
+		default:
+			s.lock.RUnlock()
+		}
 	}
+}
 
-	req.GetInitial().AcceptJobTypes = acceptJobTypes
+func (s *Scheduler) connSchedulerHosts(ctx context.Context) (schedulerv1pb.Scheduler_WatchHostsClient, []string, error) {
+	log.Infof("Attempting to connect to scheduler: %s", s.addresses[0])
 
-	clients, err := m.clients.All(ctx)
+	cl, err := client.New(ctx, s.addresses[0], s.security)
```

Comment:
Can you add a comment for future reference that mentions this LB under the hood due to the clusterIP?